examples/chainrules.jl

```diff
@@ -0,0 +1,156 @@
+    )
+    
+    optimize!(model)
+    @assert termination_status(model) == MOI.OPTIMAL
+    # converting to dense matrix
+    return JuMP.value.(p.data)
+end
+
+@show unit_commitment([1.0, 1.2, 1.4, 1.6], [1.0, 1.2, 1.4, 1.6], [1000.0, 1500.0], [500.0, 1000.0])
+
+# Forward differentiation rule for the solution map of the unit commitment problem
+# taking in input perturbations on the input parameters and returning perturbations propagated to the result
+function ChainRulesCore.frule((_, Δload1_demand, Δload2_demand, Δgen_costs, Δnoload_costs), ::typeof(unit_commitment), load1_demand, load2_demand, gen_costs, noload_costs)
+    # creating the UC model with a DiffOpt optimizer wrapper around Clp
+    model = Model(() -> diff_optimizer(Clp.Optimizer))
+    # building and solving the main model
+    pv = unit_commitment(load1_demand, load2_demand, gen_costs, noload_costs, model=model)
+    energy_balance_cons = model[:energy_balance_cons]
+
+    # Setting some perturbation of the right-hand side of the energy balance constraints
+    # the RHS is equal to the sum of load demands at each period.
+    # the corresponding perturbation are set accordingly as the set of perturbations of the two loads
+    MOI.set.(
+        model,
+        DiffOpt.ForwardIn{DiffOpt.ConstraintConstant}(), energy_balance_cons,
+        [d1 + d2 for (d1, d2) in zip(Δload1_demand, Δload1_demand)],
+    )
+
+    p = model[:p]
+    u = model[:u]
+
+    # setting the perturbation of the linear objective
+    for t in size(p, 2)
+        MOI.set(model, DiffOpt.ForwardIn{DiffOpt.LinearObjective}(), p[1,t], Δgen_costs[1])
+        MOI.set(model, DiffOpt.ForwardIn{DiffOpt.LinearObjective}(), p[2,t], Δgen_costs[2])
+        MOI.set(model, DiffOpt.ForwardIn{DiffOpt.LinearObjective}(), u[1,t], Δnoload_costs[1])
+        MOI.set(model, DiffOpt.ForwardIn{DiffOpt.LinearObjective}(), u[2,t], Δnoload_costs[2])
+    end
+    DiffOpt.forward(JuMP.backend(model))
+    # querying the corresponding perturbation of the decision
+    Δp = MOI.get.(model, DiffOpt.ForwardOut{MOI.VariablePrimal}(), p)
+    return (pv, Δp.data)
+end
+
+
+load1_demand = [1.0, 1.2, 1.4, 1.6]
+load2_demand = [1.0, 1.2, 1.4, 1.6]
+gen_costs = [1000.0, 1500.0]
+noload_costs = [500.0, 1000.0]
+
+Δload1_demand = 0 * load1_demand .+ 0.1
+Δload2_demand = 0 * load2_demand .+ 0.2
+Δgen_costs = 0 * gen_costs .+ 0.1
+Δnoload_costs = 0 * noload_costs .+ 0.4
+@show (pv, Δpv) = ChainRulesCore.frule((nothing, Δload1_demand, Δload2_demand, Δgen_costs, Δnoload_costs), unit_commitment, load1_demand, load2_demand, gen_costs, noload_costs)
+
+# Reverse-mode differentiation of the solution map
+# The computed pullback takes a seed for the optimal solution `̄p` and returns
+# derivatives wrt each input parameter.
+function ChainRulesCore.rrule(::typeof(unit_commitment), load1_demand, load2_demand, gen_costs, noload_costs; model = Model(() -> diff_optimizer(Clp.Optimizer)))
```

Comment:
we should write this assuming the Model they use isn't a `diff_optimizer` model.
And we should unwrap and rewrap the model

Reply:
unwrapping and rewrapping the model results in the same issue as above